### PR TITLE
Copy webpack plugin integration

### PR DESCRIPTION
**Pull request by davidmpaz**
Hi,

first of all, thanks for this project. I am writing to ask whether could be possible to include this patch in main stream or in case is not good, lets try to find/coordinate a solution for include copied asset in manifest.

My use case is that I have several images used in php templates, these are not compiled/processed by Webpack and as such they are never copied to target directory. 

In this scenario, plugins like [copy-webpack-plugin](https://github.com/kevlened/copy-webpack-plugin) are very helpful. One missing feature, I think many people will appreciate, is to have those copied assets in the manifest file also.

Some notes on my findings:

* variable [moduleAssets](https://github.com/danethurber/webpack-manifest-plugin/blob/v1.1.0/lib/plugin.js#L67) never contain copied assets, making check on `name` variable fail.
* Still we have assets in `compilation.getStats().toJson().assets`.
* In case you ask why to use plugin instead a file-loader. I can say I tried, but the solution I was coming into was not good to me. I needed to mix build logic into code that is run inside application etc..

Many thanks in advance and best regards.

### CODE REVIEW

**Comment by weaverryan** [comment]
@davidmpaz could you add a test for this? Does it work in all the situations you've found? Or are there some caveats?

**Comment by mastilver** [comment]
I'm sorry you need to find another solution for that, it will output 'normal' file twice

I'm also not sure about the option, ideally I would like to land this on `v2` turned on by default

**Comment by mastilver** [comment]
@davidmpaz If you test again the latest develop and `includeCopiedAssets` tests will fail

**Comment by davidmpaz** [comment]
Hi @mastilver precisely today I was trying to add some tests and figure it out of that duplication with tests. I should have had added tests since the beginning. Test are failing in travis now but I will fix them soon. Today I can not continue.

@weaverryan I don't see any other caveats besides the issue with duplication mentioned, which should be fixed now.

**Comment by mastilver** [comment]
@davidmpaz No problem at all :)

Take your time, I'm planning to merge your PR for `v2` which is not happening anytime soon
I will review your existing work. Pick it up only when you are ready ;)

File: lib/plugin.js [hunk 1]
```diff
@@ -63,10 +64,15 @@ ManifestPlugin.prototype.apply = function(compiler) {
 
     // module assets don't show up in assetsByChunkName.
     // we're getting them this way;
+    var allowCopied = this.opts.includeCopiedAssets;
```
Comment (mastilver) [inline lib/plugin.js]: Remove the option, it will added as a breaking change, so no need to be backward compatible

File: spec/plugin.spec.js [hunk 2]
```diff
@@ -221,4 +221,48 @@ describe('ManifestPlugin', function() {
       });
     });
   });
+
+  describe('assets added by third parties to "compilation.assets"', function() {
+    it('are included also when enabled.', function(done) {
+
+      var options = webpackConfig({
```
Comment (mastilver) [inline spec/plugin.spec.js]: Can you not use `webpackConfig`?
Follow the same pattern as `with ExtractTextPlugin`

File: spec/plugin.spec.js [hunk 3]
```diff
@@ -221,4 +221,48 @@ describe('ManifestPlugin', function() {
...
+        manifestOptions: {
+            includeCopiedAssets: true
+        }
+      });
+
+      // add plugin to get third party assets included
+      let Mock = require(path.join(__dirname, './plugin-mock'));
+      options.plugins.unshift(new Mock());
+
+      webpackCompile(options, function(manifest, stats) {
```
Comment (mastilver) [inline spec/plugin.spec.js]: `webpackCompile(options, {}, function(manifest, stats) {`

File: spec/plugin.spec.js [hunk 4]
```diff
@@ -221,4 +221,48 @@ describe('ManifestPlugin', function() {
...
+        expect(manifest['third.party.js']).toEqual('third.party.js');
```
Comment (mastilver) [inline spec/plugin.spec.js]: Can you follow this pattern: `expect(manifest).toEqual(...)`, it will prevents future issues from slipping through

File: spec/plugin.spec.js [hunk 5]
```diff
@@ -221,4 +221,48 @@ describe('ManifestPlugin', function() {
...
+    it('are included also when enabled.', function(done) {
+
+      var options = webpackConfig({
+        manifestOptions: {
+            includeCopiedAssets: true
+        }
+      });
+
+      // add plugin to get third party assets included
+      let Mock = require(path.join(__dirname, './plugin-mock'));
```
Comment (mastilver) [inline spec/plugin.spec.js]: `let Mock = require('./plugin-mock');` and can yoou put it at the very beginning of the file (Maybe find a better name ;) )

File: spec/plugin.spec.js [hunk 6]
```diff
@@ -221,4 +221,48 @@ describe('ManifestPlugin', function() {
...
+        done();
+      });
+    });
+  });
+
+  describe('assets added by third parties to "compilation.assets"', function() {
+    it('are not included when disabled.', function(done) {
+
+      var options = webpackConfig({
+        entry: {
```
Comment (mastilver) [inline spec/plugin.spec.js]: use `context: __dirname` and `one: ''./fixtures/file.js''`

File: spec/plugin.spec.js [hunk 7]
```diff
@@ -221,4 +221,48 @@ describe('ManifestPlugin', function() {
...
+      options.plugins.unshift(new Mock());
+
+      webpackCompile(options, function(manifest, stats) {
+        expect(manifest['third.party.js']).toEqual('third.party.js');
+        done();
+      });
+    });
+  });
+
+  describe('assets added by third parties to "compilation.assets"', function() {
```
Comment (mastilver) [inline spec/plugin.spec.js]: No need for this test, as I want to make it a default

File: lib/plugin.js (hunk 1 above)
Comment (davidmpaz) [inline lib/plugin.js]: Ok. understandable.

File: spec/plugin.spec.js (hunk 2 above)
Comment (davidmpaz) [inline spec/plugin.spec.js]: Did this since could not find a way to use mock plugin only for one test. Now it is added to all, directly in initial configuration.

File: spec/plugin.spec.js (hunk 5 above)
Comment (davidmpaz) [inline spec/plugin.spec.js]: Done :)

File: spec/plugin.spec.js (hunk 3 above)
Comment (davidmpaz) [inline spec/plugin.spec.js]: Fixed

File: spec/plugin.spec.js (hunk 7 above)
Comment (davidmpaz) [inline spec/plugin.spec.js]: I would like to have a way to test it without using the plugin. Any idea how best to do. Currently I always add it.

File: spec/plugin.spec.js (hunk 6 above)
Comment (davidmpaz) [inline spec/plugin.spec.js]: does it make sense to test when no assets are added (or not found) in entry?

File: lib/plugin.js [hunk 8]
```diff
@@ -24,6 +24,11 @@ ManifestPlugin.prototype.getFileType = function(str) {
   return ext;
 };
 
+ManifestPlugin.prototype.existNameInManifest = function(name, manifest) {
```
Comment (mastilver) [inline lib/plugin.js]: :/ I'm not sure about this one... Why is that needed?

File: spec/plugin.spec.js [hunk 9]
```diff
@@ -85,7 +88,8 @@ describe('ManifestPlugin', function() {
       }, {}, function(manifest) {
         expect(manifest).toBeDefined();
         expect(manifest).toEqual({
-          'main.js': 'main.js'
+          'main.js': 'main.js',
+          'third.party.js': 'third.party.js'
```
Comment (mastilver) [inline spec/plugin.spec.js]: I'm curious where does `third.party.js` come from?

File: spec/copy-plugin-mock.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+            source: function() {
+                return compiledMock;
+            }
+        };
+
+        callback();
+    });
+};
+
+module.exports = FakeCopyWebpackPlugin;
```
Comment (davidmpaz) [inline spec/copy-plugin-mock.js]: thanks for the suggestion! I have added a new tool to my belt :) ... Applied the formatting to relevant files.

File: lib/plugin.js (hunk 8 above)
Comment (davidmpaz) [inline lib/plugin.js]: It is not enough to test whether we have the entry in manifest keys, values need checking too. 

As far i see, we have assets in [chunk.files](https://github.com/danethurber/webpack-manifest-plugin/blob/v1.1.2/lib/plugin.js#L50) and in [stats.assets](https://github.com/danethurber/webpack-manifest-plugin/blob/v1.1.2/lib/plugin.js#L66) storing processed ones in `moduleAssets` and `manifest`, in `moduleAssets` we receive nothing, which was initial motivation of this PR, then in `manifest` i tested first only keys for avoiding duplication, it results we need more.

When using hashed values in assets name, key testing is not enough, I need to check whether file name is not in value also. For example for tests using hashes and base path, i got manifests like:
```
{ 'one.1067da9bcb5b3a357a83.js': 'one.1067da9bcb5b3a357a83.js',
  'one.js': 'one.1067da9bcb5b3a357a83.js',
  'third.party.js': 'third.party.js' }
```
the first entry in that object is wrong.

Comment (mastilver) [inline lib/plugin.js]: Oh... I see

File: spec/plugin.spec.js (hunk 9 above)
Comment (davidmpaz) [inline spec/plugin.spec.js]: The purpose of this PR is to allow any third party plugin to add assets to manifest using `compilation.assets` from webpack. That is what [copy-webpack-plugin does here](https://github.com/kevlened/copy-webpack-plugin/blob/v4.0.1/src/writeFile.js#L84)

This mean that if we execute such plugins before manifest plugin executes, then manifest plugin will be able to pick those assets and add them to final output. In this case hashing i think must rely on those external plugins since files are not processed by webpack at all.

to answer the real question: `third.party.js` simulates such behaviour, the file is added to compilation* by some plugin (`copy-plugin-mock.js`) not by webpack itself when requiring/importing assets as normal.

I don't know whether i missed some edge case, so please if you think something else must be taken into account, a bit more explanation will be welcome. I have based this on finding through debugging, not through deep knowledge of webpack :)

File: lib/plugin.js (hunk 8 above)
Comment (mastilver) [inline lib/plugin.js]: What do you think about doing this instead:

```js
else if (asset.chunks.length === 0){
  memo[asset.name] = asset.name;
}
```

File: spec/plugin.spec.js (hunk 9 above)
Comment (davidmpaz) [inline spec/plugin.spec.js]: One of the reason I tried to use first `webpackConfig` was to be able to run the specs with/without adding the mocking plugin. If you have an idea how could this be added (testing with and without using plugin), please give me some hints and I can added the code for it. Right now is always used in specs, i think it would be a nice to have in tests.

File: lib/plugin.js (hunk 8 above)
Comment (davidmpaz) [inline lib/plugin.js]: Please could you explain a bit more. I can not follow the idea.

File: spec/plugin.spec.js (hunk 9 above)
Comment (mastilver) [inline spec/plugin.spec.js]: Oh I see, I didn't realised you added it to all tests

Comment (mastilver) [inline spec/plugin.spec.js]: Can you create a new test that follow the same logic as `with ExtractTextPlugin` but for `FakeCopyWebpackPlugin`

File: spec/copy-plugin-mock.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
+function FakeCopyWebpackPlugin() {
```
Comment (mastilver) [inline spec/copy-plugin-mock.js]: Move this file to `helpers/copy-plugin-mock.js`

File: lib/plugin.js [hunk 12]
```diff
@@ -14,7 +14,7 @@ function ManifestPlugin(opts) {
   }, opts || {});
 }
 
-ManifestPlugin.prototype.getFileType = function(str) {
+ManifestPlugin.prototype.getFileType = function (str) {
```
Comment (mastilver) [inline lib/plugin.js]: leave it like this `function(str)`

File: spec/plugin.spec.js (hunk 9 above)
Comment (davidmpaz) [inline spec/plugin.spec.js]: Ok i will do. I have time now.

File: spec/copy-plugin-mock.js (hunk 11 above)
Comment (davidmpaz) [inline spec/copy-plugin-mock.js]: ok.

File: lib/plugin.js (hunk 12 above)
Comment (davidmpaz) [inline lib/plugin.js]: uhmm. i think i can not trust on idea IDEs for formatting using editorconfig rules. I will look on editorconfig for a rule to add such restriction. Otherwise other will incur in same error

File: README.md [hunk 13]
```diff
@@ -53,3 +53,4 @@ new ManifestPlugin({
 * `stripSrc`: removes unwanted strings from source filenames (string or regexp)
 * `writeToFileEmit`: If set to `true` will emit to build folder and memory in combination with `webpack-dev-server`
 * `cache`: In [multi-compiler mode](https://github.com/webpack/webpack/tree/master/examples/multi-compiler) webpack will overwrite the manifest on each compilation. Passing a shared `{}` as the `cache` option into each compilation's ManifestPlugin will combine the manifest between compilations.
+* `includeCopiedAssets`: Allow to include assets that where not processed by webpack but included into assets by plugins like [copy-webpack-plugin](https://github.com/kevlened/copy-webpack-plugin)
```
Comment (mastilver) [inline README.md]: remove this

File: lib/plugin.js (hunk 8 above)
Comment (mastilver) [inline lib/plugin.js]: Rather than checking if a value is already present in the manifest (Which I can see could conflict with #55 , just make sure that the asset is a real asset, not something generated by webpack (make sure that `asset.chunks` is empty). Not sure if I make sense?

File: spec/plugin.spec.js [hunk 14]
```diff
@@ -6,6 +6,8 @@ var _ = require('lodash');
 var ExtractTextPlugin = require('extract-text-webpack-plugin');
 var extractTextPluginMajorVersion = require('extract-text-webpack-plugin/package.json').version.split('.')[0];
 var plugin = require('../index.js');
+// add plugin to get third party assets included
+var FakeCopyWebpackPlugin = require(path.join(__dirname, './helpers/copy-plugin-mock'));
```
Comment (mastilver) [inline spec/plugin.spec.js]: `var FakeCopyWebpackPlugin = require('./helpers/copy-plugin-mock');`

File: lib/plugin.js (hunk 8 above)
Comment (davidmpaz) [inline lib/plugin.js]: To be honest, this is the point in which my knowledge of webpack start shortening. I am not sure if it makes sense neither.

Something i did notice when debugging was that files added by copy plugin were never coming in `compilation.chunks`, but in `compilation.stats.assets`. Does this answer your question?

Unfortunately I could not extract much knowledge from issue #55 too many use cases i don't know at all. Which part do you think will conflict?

**Comment by mastilver** [comment]
Hi @davidmpaz 

Can you let me know if you have time to carry on with this one, I want to release `2.x` next week-end

Can you remove all your style changes: no space between function and parentheses 

Thank you :)

**Comment by davidmpaz** [comment]
Hi @mastilver,

sure. I did notice that you merged to update branches but thought you were going to continue working on it.

I will take a look to finish it between today and tomorrow, otherwise I will not have time later :)

cheers

**Comment by mastilver** [comment]
Great, thank you@ :)

Let me know if you can't so I can carry on

One last thing, for those files, can you see `isModuleAsset` to `false`, please, that way there is a way to tell them apart

**Comment by davidmpaz** [comment]
Hi @mastilver,

after looking better to current work I decided would be easier to just create another pull request than re-basing this one. The changes have been to much to conciliate again what was there :)

Please lets continue in another PR I will made now.
